### PR TITLE
Overriding parameters with settings

### PR DESCRIPTION
**Pull request by cpelley**
Overriding Node parameters with node attributes (settings).
Exposed this requirement as part of my tackling https://github.com/MetOffice/epp_workflows/pull/66
Technically, this is non-blocking as IMPROVER or EPP should be responsible for generating the networkx graph.  Dagrunner ExecuteGraph simply provides the means to pass it nodes and edges to do this for you.

Additional minor changes:
- Removed `os.getpid()` from the function signature which changed the documentation build every time.
- Added `node_properties` argument to `Input` plugin to make consistent with other node aware plugins.

## Why do we want this?

More easily explained with an example (illustrative).  precipitation accumulation is defined with a period in its diagnostic name.  This is a function of the leadtime (x, y here).  We don't want to unnecessarily be defining our node identity with a period in this diagnostic name or through adding a period parameter, since the period is only characteristic of the output filename, not its identity (the leadtime is enough to characterise its identity and we don't want to be deducing the period for a node everywhere we want to connect to it).
Instead, what we want is to override the diagnostic property using a node attribute (setting) for the node where it matters (saving).  That is, when saving to disk, as the filepath will be derived using this period in the name.

```python
diagnostic_name = "precipitation_accumulation"
...
for leadtime in [x, y]:
    NODES.append(
        [
            Node(model="gl", diagnostic="precipitation_accumulation", step=..., leadtime=leadtime),
            Node(model="gl", diagnostic="precipitation_accumulation", step="save", leadtime=leadtime),
        ]
    )

    period = PERIOD_LOOKUP[leadtime/HOUR]
SETTINGS[Node(model="gl", diagnostic="precipitation_accumulation", step="save", leadtime=leadtime)] = {
        "call": {...}, "diagnostic": f"{diagnostic_name}{period}"
    }
...
```

## Issues

- https://github.com/MetOffice/improver_suite/issues/2253
- https://github.com/MetOffice/epp_workflows/pull/55

### CODE REVIEW

File: dagrunner/execute_graph.py [hunk 1]
```diff
@@ -235,7 +235,7 @@ def _get_networkx(networkx_graph):
     else:
         try:
             edges, nodes = networkx_graph
-            nodes = {k: nodes[k] | _process_nodes(k) for k in nodes.keys()}.items()
+            nodes = {k: _process_nodes(k) | nodes[k] for k in nodes.keys()}.items()
```
Comment (cpelley) [inline dagrunner/execute_graph.py:238]: This is the simple change of this PR... (changing the order of the precedence to allowing node attributes to override Node properties.

File: dagrunner/plugin_framework.py [hunk 2]
```diff
@@ -239,7 +239,7 @@ def __call__(
 
 
 class Input(NodeAwarePlugin):
-    def __call__(self, filepath, **kwargs):
+    def __call__(self, filepath, node_properties=None, **kwargs):
```
Comment (cpelley) [inline dagrunner/plugin_framework.py:242]: minor change, making node properties explicitly an argument to make consistent with other NodeAwarePlugins

File: dagrunner/utils/__init__.py [hunk 3]
```diff
@@ -33,7 +33,7 @@ def process_path(fpath: str) -> str:
     return fpath
 
 
-def get_proc_mem_stat(pid=os.getpid()):
+def get_proc_mem_stat(pid=None):
```
Comment (cpelley) [inline dagrunner/utils/__init__.py:36]: this ensures that the reference documentation no longer has the pid number encoded in it (changing the documentation each time you build it).

File: docs/dagrunner.utils.md [hunk 4]
```diff
@@ -382,7 +382,7 @@ Generate an argparse from a function signature
 ### Call Signature:
 
 ```python
-get_proc_mem_stat(pid=2356)
+get_proc_mem_stat(pid=None)
```
Comment (cpelley) [inline docs/dagrunner.utils.md:385]: pid number no longer shows in the ref. docs

File: dagrunner/plugin_framework.py [hunk 5]
```diff
@@ -248,7 +248,8 @@ def __call__(self, filepath, **kwargs):
         `NodeAwarePlugin`.
 
         Args:
-        - filepath (str): The filepath to be expanded.
+        - `filepath` (str): The filepath to be expanded.
+        - `node_properties`: node properties passed by the plugin executor.
         - **kwargs: Keyword arguments to be used in the expansion.  Node
           properties/attributes are additionally included here as a node aware plugin.
```
Comment (mo-robert-purvis) [inline dagrunner/plugin_framework.py:254]: Comment needs changing now that node properties have their own explicit argument.

**Review by mo-robert-purvis — APPROVED** [review]

**Comment by SamGriffithsMO** [comment]
I thought overriding was something we were actively trying to remove when going to dagrunner, won't this increase the scope to reintroduce mass overriding again?

**Comment by cpelley** [comment]
> I thought overriding was something we were actively trying to remove when going to dagrunner, won't this increase the scope to reintroduce mass overriding again?

The PR title could be somewhat misleading.  This is to do with simultaneously supporting node parameters for its identity while retaining support for a node attribute that doesn't impact this identity (the node parameter remains the same for node identity, not overridden).
The overriding behaviour that I have mention previously concerned more how settings are defined with no way to understand what they are besides debugging the graph.

**Pull request merged**